pinot-core/src/main/java/org/apache/pinot/core/operator/dociditerators/LuceneIndexScanDocIdIterator.java

```diff
@@ -0,0 +1,90 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.core.operator.dociditerators;
+
+import org.apache.lucene.document.Document;
+import org.apache.lucene.search.IndexSearcher;
+import org.apache.pinot.core.common.Constants;
+import org.apache.pinot.core.segment.creator.impl.V1Constants;
+import org.apache.pinot.core.segment.index.readers.text.LuceneIndexSearcherReferenceManager;
+import org.apache.pinot.core.segment.index.readers.text.LuceneTextIndexReader;
+import org.roaringbitmap.IntIterator;
+
+
+public class LuceneIndexScanDocIdIterator implements IndexBasedDocIdIterator {
+
+  private int _currentDocId = -1;
+  private int _startDocId;
+  private int _endDocId;
+  private final IntIterator _docIDIterator;
+  private final IndexSearcher _indexSearcher;
+  private final LuceneIndexSearcherReferenceManager _searcherReferenceManager;
+  private final int _numDocs;
+
+  public LuceneIndexScanDocIdIterator(LuceneTextIndexReader.LuceneSearchResult luceneSearchResult) {
+    _docIDIterator = luceneSearchResult.getDocIDs().getIntIterator();
```

Comment:
why not just get the bitmap and reused all the iterators on top of bitmap

Reply:
When Lucene returns search result... we get the docIDs collected by the collector in MutableRoaringBitmap.

These are essentially Lucene internal docIDs (during index building, lucene creates separate index structures (it calls them segments) and they are merged as per its internal merge criteria so we can't rely on its internal docIds). In some cases, they can be same as Pinot docIDs but based on the index size, how many index segments it creates and how it merges etc can vary so relying on them permanently is very likely to break our code and they also advise against it.

So the way to get the actual Pinot docIDs is to store them with each Lucene document

Once the IndexSearcher finishes searching each document and all lucene docIDs are collected by the collector in the bitmap, we can do searcher.doc(lucene doc id), this will get the document and we can retrieve the actual docID from the document.

Now why can't we do searcher.doc(lucene doc id) as part of collector? It suffers severe performance -- (something there documentation also says not to do this while the collector is being invoked).

So that's why this bitmap with lucene docids is then iterated by LuceneIndexDocIdIterator and for each lucene doc ID, we get the actual docID and this is what is returned by the iterator for rest of the filter processing to continue. Once we have iterated completely, we also release the searcher via reference manager (NO-OP for offline search).

Now if we were to use BitMapDocIdSet -- which directly works against the bitmap of docIDs, we have two options

(1) Change the implementation of BitmapDocIdSet to take LuceneSearchResult in the constructor and do all of this there and in BitMapDocIdIterator
OR
(2) Once the call to textIndexReader.getDocIds(predicate) comes back into TextMatchFilterOperator, we iterate over the lucene docIDs there itself and get the Pinot docIDs (as described above) and collect them in a new bitmap. Pass this bitmap over to BitmapDocIdSet. Everything will work correctly but then we have to construct another bitmap

Reply:
Discussed offline. It is better to iterate over the lucene doc id set upfront and build the bitmap with actual pinot doc ids and then have rest of the filter processing work off that bimap

Reply:
As a follow-up we can optimize this path to see if a construction of new bitmap is needed or not